### ISSUE TITLE
Understanding FastGradientMethod usage and how to force output label to desired value

### ISSUE THREAD

**Issue by ftarlao** [tensorflow/cleverhans#589]
I have a keras model (CNN with final softmax) that is an RGB image classifier that labels input images in 5 categories (one-hot encoded).
A simplified version of my code is the following:
```
wrap = KerasModelWrapper(model)
fgsm = FastGradientMethod(wrap, sess=session)
fgsm_params = {'eps': 16. / 256,
               'clip_min': 0.,
               'clip_max': 1.
               }
x = tf.placeholder(tf.float32, shape=(None, img_rows, img_cols,
                                      nchannels))
adv_x = fgsm.generate(x, **fgsm_params)
#original image is a tensor containing only one RGB image, shape=(1,48,48,3) 
adv_image = adv_x.eval(session=session, feed_dict={x: original_image})
```
**Question 1**
From my understanding, 'eps' is the input variation step (minimum change).
I have noticed that the final outcome is highly affected by _eps_, sometimes I need high _eps_ in order to obtain an effective adversarial image (having an image O, with label l<sub>O</sub>  sometimes FGM fails to produce adversarial image O' with l<sub>O'</sub>!= l<sub>O</sub>, e.g., for  l<sub>O</sub> = [0,0,1,0,0] we still obtain l<sub>O'</sub> = [0,0,1,0,0], failing to generate an adversarial image with different label.

Does FGM always find out a working adversarial image? Is it normal that it fails? Is there a way to obtain an estimated quality of the generated adversarial image (without predicting with model)? Why is the _eps_ step so important? Is there a way to tell FGM to try harder searching for the adversarial image(e.g, more steps)?

**Question 2**
I have also experimented _y_ and _y_target params_. Can you also explain me what are the params `'y', 'y_target'`? I thought  `'y_target'` tells that we want to generate an adversarial image that targets a specific category. For example I thought that `y_target = [[0,1,0,0,0]]` in `feed_dict` should force to generate an adversarial image which is classified with the 2th class from the model.
Am I right?
..or do I miss something?
P.s: my problem is that setting `y_target` fails to produce adversarial images.

please give me few tips.. 
Regards

**Comment by ftarlao** [comment]
In case you prefer, or you like to level up on SO, I have also posted the question here:

https://stackoverflow.com/questions/52501833/doubts-with-cleverhans-fastgradientmethod-fgm-adversarial-image-generation

**Comment by npapernot** [comment]
Q1: 

* FGSM (like any attack) is not guaranteed to find an adversarial image that is misclassified by the model because it makes approximations when solving the optimization problem that defines an adversarial example. The attack can fail to find adversarial images for various reasons, one common reason is gradient masking. You can read about it [in this blog post](http://www.cleverhans.io/security/privacy/ml/2017/02/15/why-attacking-machine-learning-is-easier-than-defending-it.html) and in this [paper](https://arxiv.org/abs/1602.02697) as well as this [paper](https://arxiv.org/abs/1802.00420)
* The eps step is important because it is the magnitude of the perturbation. The attack first computes the direction in which to perturb the image (using gradients of the model) and then takes a step of size eps in that direction. Hence, eps corresponds roughly to what intuitively one would think of the "power" of the attack.
* You can find a multi-step variant of FGSM in `BasicIterativeMethod`

Q2: y is used to specify labels in the case of an untargeted attack (any wrong class is considered as success for the adversary) whereas y_target is used to specify a target class in the targeted attack case (the adversary is successful only if the model makes a particular misclassification in a chosen class). It is often the case that targeted attacks require more perturbation (i.e., higher eps values in the FGSM case) than untargeted attacks.

**Comment by ftarlao** [comment]
I copied your answer to SO question. Regards
